v2/apiserver/internal/api/jobs.go

```diff
@@ -127,6 +126,30 @@ type JobSpec struct {
 	Host *JobHost `json:"host,omitempty" bson:"host,omitempty"`
 }
 
+func (js JobSpec) EqualTo(js2 JobSpec) bool {
+	// Compare SidecarContainers maps
+	if len(js.SidecarContainers) != len(js2.SidecarContainers) {
+		return false
+	}
+	for name, sidecarContainer := range js.SidecarContainers {
+		if !sidecarContainer.EqualTo(js2.SidecarContainers[name]) {
+			return false
+		}
+	}
+	// Need to check if Host values are nil
+	if js.Host == nil || js2.Host == nil {
+		if js.Host == nil && js2.Host == nil {
+			return js.PrimaryContainer.EqualTo(js2.PrimaryContainer) &&
+				js.TimeoutDuration == js2.TimeoutDuration
+		}
+		return false
+	}
+	// Return remaining field equivalence
+	return js.PrimaryContainer.EqualTo(js2.PrimaryContainer) &&
+		js.TimeoutDuration == js2.TimeoutDuration &&
+		js.Host.EqualTo(js2.Host)
```

Comment:
This is the part I worry about...

If we add new fields in the future, it will be easy to forget to amend this bit of logic and I don't think we can necessarily count on the tests to fail in that case. So we could end up regressing on the bug this is meant to fix.

Is there some possibility that instead of comparing specific fields here by name, that we could nil out the fields we already evaluated for equality and use reflection to test equality of what's left?

Reply:
That's a good idea.  I updated these methods to do so.  Let me know how it looks...